R/ctstm.R

```diff
@@ -247,6 +251,12 @@ IndivCtstmTrans <- R6::R6Class(
     #' `clock = "mix"`.
     reset_states = NULL,
     
+    #' @field reset_transitions A vector denoting the transitions in which time resets. 
+    #' Hazard functions are always a function of elapsed time since either the 
+    #' start of the model or from when time was previously reset. Only used if 
```

Comment:
This code from `indiv-ctstm.h` seems to suggest time "reset" in a different way then with reset states.

```
else if (clock_ == "mixt"){
          if (is_reset_transition(trans_ids[i])){
            random_times[i] = transmod_->random(trans_ids[i], sample); 
          } else{
           random_times[i] = transmod_->trandom(trans_ids[i], sample, time_) - time_; 
          }
      }
```

 I think the documentation should reflect this.

Reply:
Devin: thank you for the feedback.

For `clock="mix"`, your code for a "forward" state is different to that for `clock="forward"` (unless all states are forward). The variable `clockmix_time_` seems to sum the time from a reset state -- which is a very interesting way to represent one or more interventions. It took me a while to understand your implementation:).

Some possible time scales of interest:

1. Reset time / time in state / semi-Markov (as per (a) `clock="reset"` or (b) `clock="mix"` and a reset state or (c) `clock="mixt"` for a reset transition)
2. Time from last reset state (as per `clock="mix"` for forward states)
3. Time from study entry / Markov (as per (a) `clock="forward"` or (b) `clock="mixt"` and not being a reset transition)
4. Attained age (using `age_`)
5. Calendar period (using `time_` and a time offset)

Scales 3-5 are all based on a wall clock/calendar with a time/date offset (time at study entry, date of birth and 0 CE, respectively). **Is there a clever way to specify the transitions with different time offsets?** 

I am particularly interested in mixing transitions for attained age (for background mortality rates), time in state and time since study entry (e.g. time since cancer diagnosis). I am considering some possible changes to the `clock="mixt"` implementation:

-  remove the `reset_transitions` argument and add a `transition_type` argument which takes a vector of strings for the type of transitions; this could include `"reset"`, `"age"` and `"time"` (allowing for 1, 3 and 4). Code:

```
       if (clock_ == "mixt"){
          if (transition_type[i] == "reset"){
            random_times[i] = transmod_->random(trans_ids[i], sample); 
          } else if (transition_type[i] == "age"){
            random_times[i] = transmod_->trandom(trans_ids[i], sample, age_) - age_; 
          } else{ 
           random_times[i] = transmod_->trandom(trans_ids[i], sample, time_) - time_; 
          }
      }
```

- alternatively, keep the `reset_transitions` argument and add a `time_offset` argument (allowing for 1, 3, 4 and 5). This may require a time-offset be specified for all transitions. Code:

```
       if (clock_ == "mixt"){
          if (transition_type[i] == "reset"){
            random_times[i] = transmod_->random(trans_ids[i], sample); 
          } else{ 
            random_times[i] = transmod_->trandom(trans_ids[i], sample, time_ + time_offset[i]) - (time_ + time_offset[i]); 
          }
```

**What do you think about either of these approaches?**

Finally, is there a relatively simple approach which would allow for all five time scales?

Reply:
Devin: I have now implemented the first approach.

It would still be useful to know if there is a clever way to specify the transitions with different time offsets.

Reply:
I agree with the chosen approach. "reset", "age", and "time" are the easy options for users. We can consider adding a `time_offset` in the future, but its seems it would be more complicated since it could not only vary by transition, but also (I think) by individual. 

If you are also okay with it as well, I have a few suggestions on the implementation.

Reply:
Agreed.